// @beyond-js/kernel/error/error.sjs.js
System.register(["@beyond-js/kernel/bundle"], function (_export, _context) {
  "use strict";

  var dependency_0, __Bundle, __pkg, ims, __beyond_pkg, hmr;
  return {
    setters: [function (_beyondJsKernelBundle) {
      dependency_0 = _beyondJsKernelBundle;
    }],
    execute: function () {
      ({
        Bundle: __Bundle
      } = dependency_0);
      __pkg = new __Bundle({
        "module": {
          "vspecifier": "@beyond-js/kernel@0.1.9/error"
        },
        "type": "ts"
      }, _context.meta.url).package();
      ;
      __pkg.dependencies.update([]);
      ims = new Map();
      /***********************
      INTERNAL MODULE: ./error
      ***********************/
      ims.set('./error', {
        hash: 1344789388,
        creator: function (require, exports) {
          "use strict";

          const prepareStackTrace = Error => {
            Error.prepareStackTrace = (err, frames) => {
              for (const frame of frames) {
                if (frame.isNative()) continue;
                const file = frame.getFileName();
                const line = frame.getLineNumber();
                const column = frame.getColumnNumber();
                console.log(file, line, column);
              }
              return err.stack;
            };
          };
          Error.prepareStackTrace && prepareStackTrace(Error);
        }
      });

      // Module exports
      __pkg.exports.process = function ({
        require,
        prop,
        value
      }) {};
      _export("__beyond_pkg", __beyond_pkg = __pkg);
      _export("hmr", hmr = new function () {
        this.on = (event, listener) => void 0;
        this.off = (event, listener) => void 0;
      }());
      __pkg.initialise(ims);
    }
  };
});
