// @beyond-js/kernel/error/error.cjs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hmr = exports.__beyond_pkg = void 0;
var dependency_0 = require("@beyond-js/kernel/bundle");
const {
  Bundle: __Bundle
} = dependency_0;
const __pkg = new __Bundle({
  "module": {
    "vspecifier": "@beyond-js/kernel@0.1.9/error"
  },
  "type": "ts"
}).package();
;
__pkg.dependencies.update([]);
const ims = new Map();

/***********************
INTERNAL MODULE: ./error
***********************/

ims.set('./error', {
  hash: 1344789388,
  creator: function (require, exports) {
    "use strict";

    const prepareStackTrace = Error => {
      Error.prepareStackTrace = (err, frames) => {
        for (const frame of frames) {
          if (frame.isNative()) continue;
          const file = frame.getFileName();
          const line = frame.getLineNumber();
          const column = frame.getColumnNumber();
          console.log(file, line, column);
        }
        return err.stack;
      };
    };
    Error.prepareStackTrace && prepareStackTrace(Error);
  }
});

// Module exports
__pkg.exports.process = function ({
  require,
  prop,
  value
}) {};
const __beyond_pkg = __pkg;
exports.__beyond_pkg = __beyond_pkg;
const hmr = new function () {
  this.on = (event, listener) => void 0;
  this.off = (event, listener) => void 0;
}();
exports.hmr = hmr;
__pkg.initialise(ims);
