// @beyond-js/react-18-widgets/page/page.browser.mjs
import * as dependency_0 from '@beyond-js/kernel/bundle';
import * as dependency_1 from '@beyond-js/react-18-widgets/base';
import * as dependency_2 from '@beyond-js/widgets/routing';

const {Bundle: __Bundle} = dependency_0;
const __pkg = new __Bundle({"module":{"vspecifier":"@beyond-js/react-18-widgets@0.0.4/page"},"type":"ts"}, import.meta.url).package();;

__pkg.dependencies.update([['@beyond-js/react-18-widgets/base', dependency_1],['@beyond-js/widgets/routing', dependency_2]]);

const ims = new Map();

/**********************
INTERNAL MODULE: ./page
**********************/

ims.set('./page', {hash: 2113815253, creator: function (require, exports) {
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PageReactWidgetController = void 0;
var _base = require("@beyond-js/react-18-widgets/base");
var _routing = require("@beyond-js/widgets/routing");
/*bundle*/class PageReactWidgetController extends _base.ReactWidgetController {
  #uri;
  get uri() {
    return this.#uri;
  }
  mount() {
    return super.mount({
      uri: this.#uri
    });
  }
  async initialise() {
    this.#uri = new _routing.PageURI({
      widget: this.widget
    });
    await super.initialise();
  }
}
exports.PageReactWidgetController = PageReactWidgetController;
}});

__pkg.exports.descriptor = [{"im":"./page","from":"PageReactWidgetController","name":"PageReactWidgetController"}];

export let PageReactWidgetController;

// Module exports
__pkg.exports.process = function({require, prop, value}) {
    (require || prop === 'PageReactWidgetController') && (PageReactWidgetController = require ? require('./page').PageReactWidgetController : value);

};
export const __beyond_pkg = __pkg;

export const hmr = new (function () {
    this.on = (event, listener) => void 0;
    this.off = (event, listener) => void 0;
});


__pkg.initialise(ims);
