// @beyond-js/react-18-widgets/page/page.sjs.js
System.register(["@beyond-js/kernel/bundle", "@beyond-js/react-18-widgets/base", "@beyond-js/widgets/routing"], function (_export, _context) {
  "use strict";

  var dependency_0, dependency_1, dependency_2, __Bundle, __pkg, ims, PageReactWidgetController, __beyond_pkg, hmr;
  _export("PageReactWidgetController", void 0);
  return {
    setters: [function (_beyondJsKernelBundle) {
      dependency_0 = _beyondJsKernelBundle;
    }, function (_beyondJsReact18WidgetsBase) {
      dependency_1 = _beyondJsReact18WidgetsBase;
    }, function (_beyondJsWidgetsRouting) {
      dependency_2 = _beyondJsWidgetsRouting;
    }],
    execute: function () {
      ({
        Bundle: __Bundle
      } = dependency_0);
      __pkg = new __Bundle({
        "module": {
          "vspecifier": "@beyond-js/react-18-widgets@0.0.4/page"
        },
        "type": "ts"
      }, _context.meta.url).package();
      ;
      __pkg.dependencies.update([['@beyond-js/react-18-widgets/base', dependency_1], ['@beyond-js/widgets/routing', dependency_2]]);
      ims = new Map();
      /**********************
      INTERNAL MODULE: ./page
      **********************/
      ims.set('./page', {
        hash: 2113815253,
        creator: function (require, exports) {
          "use strict";

          Object.defineProperty(exports, "__esModule", {
            value: true
          });
          exports.PageReactWidgetController = void 0;
          var _base = require("@beyond-js/react-18-widgets/base");
          var _routing = require("@beyond-js/widgets/routing");
          /*bundle*/
          class PageReactWidgetController extends _base.ReactWidgetController {
            #uri;
            get uri() {
              return this.#uri;
            }
            mount() {
              return super.mount({
                uri: this.#uri
              });
            }
            async initialise() {
              this.#uri = new _routing.PageURI({
                widget: this.widget
              });
              await super.initialise();
            }
          }
          exports.PageReactWidgetController = PageReactWidgetController;
        }
      });
      __pkg.exports.descriptor = [{
        "im": "./page",
        "from": "PageReactWidgetController",
        "name": "PageReactWidgetController"
      }];
      // Module exports
      __pkg.exports.process = function ({
        require,
        prop,
        value
      }) {
        (require || prop === 'PageReactWidgetController') && _export("PageReactWidgetController", PageReactWidgetController = require ? require('./page').PageReactWidgetController : value);
      };
      _export("__beyond_pkg", __beyond_pkg = __pkg);
      _export("hmr", hmr = new function () {
        this.on = (event, listener) => void 0;
        this.off = (event, listener) => void 0;
      }());
      __pkg.initialise(ims);
    }
  };
});
