// @beyond-js/backend/listen/listen.mjs
import * as dependency_0 from '@beyond-js/kernel/bundle';
import * as dependency_1 from '@beyond-js/backend/server';

const {Bundle: __Bundle} = dependency_0;
const __pkg = new __Bundle({"module":{"vspecifier":"@beyond-js/backend@0.1.5/listen"},"type":"ts"}, import.meta.url).package();;

__pkg.dependencies.update([['@beyond-js/backend/server', dependency_1]]);

const ims = new Map();

/************************
INTERNAL MODULE: ./listen
************************/

ims.set('./listen', {hash: 2537689596, creator: function (require, exports) {
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listen = listen;
var _server = require("@beyond-js/backend/server");
/*bundle*/function listen(port) {
  if (typeof globalThis.__bee === 'object') {
    const specs = globalThis.__bee.specs;
    port = specs.ports.http;
  }
  if (!port) {
    console.log('Port must be specified');
    return;
  }
  new _server.BackendServer(port);
}
}});

__pkg.exports.descriptor = [{"im":"./listen","from":"listen","name":"listen"}];

export let listen;

// Module exports
__pkg.exports.process = function({require, prop, value}) {
    (require || prop === 'listen') && (listen = require ? require('./listen').listen : value);

};
export const __beyond_pkg = __pkg;

export const hmr = new (function () {
    this.on = (event, listener) => void 0;
    this.off = (event, listener) => void 0;
});


__pkg.initialise(ims);
