// @beyond-js/react-18-widgets/page/page.cjs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hmr = exports.__beyond_pkg = exports.PageReactWidgetController = void 0;
var dependency_0 = require("@beyond-js/kernel/bundle");
var dependency_1 = require("@beyond-js/react-18-widgets/base");
const {
  Bundle: __Bundle
} = dependency_0;
const __pkg = new __Bundle({
  "module": {
    "vspecifier": "@beyond-js/react-18-widgets@0.0.4/page"
  },
  "type": "ts"
}).package();
;
__pkg.dependencies.update([['@beyond-js/react-18-widgets/base', dependency_1]]);
const ims = new Map();

/**********************
INTERNAL MODULE: ./page
**********************/

ims.set('./page', {
  hash: 3718018090,
  creator: function (require, exports) {
    "use strict";

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.PageReactWidgetController = void 0;
    var _base = require("@beyond-js/react-18-widgets/base");
    /*bundle*/
    class PageReactWidgetController extends _base.ReactWidgetController {}
    exports.PageReactWidgetController = PageReactWidgetController;
  }
});
__pkg.exports.descriptor = [{
  "im": "./page",
  "from": "PageReactWidgetController",
  "name": "PageReactWidgetController"
}];
let PageReactWidgetController;

// Module exports
exports.PageReactWidgetController = PageReactWidgetController;
__pkg.exports.process = function ({
  require,
  prop,
  value
}) {
  (require || prop === 'PageReactWidgetController') && (exports.PageReactWidgetController = PageReactWidgetController = require ? require('./page').PageReactWidgetController : value);
};
const __beyond_pkg = __pkg;
exports.__beyond_pkg = __beyond_pkg;
const hmr = new function () {
  this.on = (event, listener) => void 0;
  this.off = (event, listener) => void 0;
}();
exports.hmr = hmr;
__pkg.initialise(ims);
