// @beyond-js/react-18-widgets/page/page.amd.js
define(["exports", "module", "@beyond-js/kernel/bundle", "@beyond-js/react-18-widgets/base", "@beyond-js/widgets/routing"], function (_exports, _amd_module, dependency_0, dependency_1, dependency_2) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.hmr = _exports.__beyond_pkg = _exports.PageReactWidgetController = void 0;
  const {
    Bundle: __Bundle
  } = dependency_0;
  const __pkg = new __Bundle({
    "module": {
      "vspecifier": "@beyond-js/react-18-widgets@0.0.4/page"
    },
    "type": "ts"
  }, _amd_module.uri).package();
  ;
  __pkg.dependencies.update([['@beyond-js/react-18-widgets/base', dependency_1], ['@beyond-js/widgets/routing', dependency_2]]);
  const ims = new Map();

  /**********************
  INTERNAL MODULE: ./page
  **********************/

  ims.set('./page', {
    hash: 2113815253,
    creator: function (require, exports) {
      "use strict";

      Object.defineProperty(exports, "__esModule", {
        value: true
      });
      exports.PageReactWidgetController = void 0;
      var _base = require("@beyond-js/react-18-widgets/base");
      var _routing = require("@beyond-js/widgets/routing");
      /*bundle*/
      class PageReactWidgetController extends _base.ReactWidgetController {
        #uri;
        get uri() {
          return this.#uri;
        }
        mount() {
          return super.mount({
            uri: this.#uri
          });
        }
        async initialise() {
          this.#uri = new _routing.PageURI({
            widget: this.widget
          });
          await super.initialise();
        }
      }
      exports.PageReactWidgetController = PageReactWidgetController;
    }
  });
  __pkg.exports.descriptor = [{
    "im": "./page",
    "from": "PageReactWidgetController",
    "name": "PageReactWidgetController"
  }];
  let PageReactWidgetController;

  // Module exports
  _exports.PageReactWidgetController = PageReactWidgetController;
  __pkg.exports.process = function ({
    require,
    prop,
    value
  }) {
    (require || prop === 'PageReactWidgetController') && (_exports.PageReactWidgetController = PageReactWidgetController = require ? require('./page').PageReactWidgetController : value);
  };
  const __beyond_pkg = __pkg;
  _exports.__beyond_pkg = __beyond_pkg;
  const hmr = new function () {
    this.on = (event, listener) => void 0;
    this.off = (event, listener) => void 0;
  }();
  _exports.hmr = hmr;
  __pkg.initialise(ims);
});
