// @beyond-js/react-18-widgets/base/base.cjs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hmr = exports.__beyond_pkg = exports.ReactWidgetController = void 0;
var dependency_0 = require("@beyond-js/kernel/bundle");
var dependency_1 = require("react");
var dependency_2 = require("react-dom/server");
var dependency_3 = require("@beyond-js/widgets/controller");
const {
  Bundle: __Bundle
} = dependency_0;
const __pkg = new __Bundle({
  "module": {
    "vspecifier": "@beyond-js/react-18-widgets@0.0.4/base"
  },
  "type": "ts"
}).package();
;
__pkg.dependencies.update([['react', dependency_1], ['react-dom/server', dependency_2], ['@beyond-js/widgets/controller', dependency_3]]);
const ims = new Map();

/****************************
INTERNAL MODULE: ./controller
****************************/

ims.set('./controller', {
  hash: 1686610974,
  creator: function (require, exports) {
    "use strict";

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.ReactWidgetController = void 0;
    var React = require("react");
    var _server = require("react-dom/server");
    var _controller = require("@beyond-js/widgets/controller");
    var _widget = require("./widget");
    /*bundle*/
    class ReactWidgetController extends _controller.WidgetServerController {
      render(props) {
        if (!this.Widget) {
          return {
            errors: [`Widget "${this.element}" does not export a Widget class`]
          };
        }
        try {
          const p = {
            Widget: this.Widget,
            styles: this.styles,
            props
          };
          const element = React.createElement(_widget.default, p);
          const html = (0, _server.renderToString)(element);
          return {
            html
          };
        } catch (exc) {
          console.error(exc.stack);
          return {
            errors: [exc.message]
          };
        }
      }
    }
    exports.ReactWidgetController = ReactWidgetController;
  }
});

/************************
INTERNAL MODULE: ./styles
************************/

ims.set('./styles', {
  hash: 3655782595,
  creator: function (require, exports) {
    "use strict";

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.default = _default;
    var React = require("react");
    function _default({
      styles
    }) {
      const head = [];
      styles.forEach(href => head.push(React.createElement("link", {
        key: href,
        href: href,
        rel: 'stylesheet'
      })));
      return React.createElement(React.Fragment, null, head);
    }
  }
});

/************************
INTERNAL MODULE: ./widget
************************/

ims.set('./widget', {
  hash: 2923914692,
  creator: function (require, exports) {
    "use strict";

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.default = _default;
    var React = require("react");
    var _styles = require("./styles");
    function _default({
      Widget,
      props,
      styles
    }) {
      const elements = [];
      elements.push(React.createElement(_styles.default, {
        key: "styles",
        styles: styles
      }));
      const widget = React.createElement(Widget, {
        key: "widget",
        ...props
      });
      elements.push(widget);
      return React.createElement(React.Fragment, null, elements);
    }
  }
});
__pkg.exports.descriptor = [{
  "im": "./controller",
  "from": "ReactWidgetController",
  "name": "ReactWidgetController"
}];
let ReactWidgetController;

// Module exports
exports.ReactWidgetController = ReactWidgetController;
__pkg.exports.process = function ({
  require,
  prop,
  value
}) {
  (require || prop === 'ReactWidgetController') && (exports.ReactWidgetController = ReactWidgetController = require ? require('./controller').ReactWidgetController : value);
};
const __beyond_pkg = __pkg;
exports.__beyond_pkg = __beyond_pkg;
const hmr = new function () {
  this.on = (event, listener) => void 0;
  this.off = (event, listener) => void 0;
}();
exports.hmr = hmr;
__pkg.initialise(ims);
